File: file-transfer-web/src/main/java/com/filetransfer/web/service/FileExtensionService.java
package com.filetransfer.web.service;

import com.filetransfer.web.entity.FileType;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Service for file extension validation, categorization, and utility operations
 */
@Service
public class FileExtensionService {
    
    // Common file extension categories
    private static final Map<String, Set<String>> EXTENSION_CATEGORIES = Map.of(
        "Text Files", Set.of(".txt", ".dat", ".log", ".csv", ".tsv"),
        "Data Files", Set.of(".json", ".xml", ".yaml", ".yml", ".properties"),
        "Archive Files", Set.of(".zip", ".gz", ".tar", ".bz2", ".xz", ".7z"),
        "Document Files", Set.of(".pdf", ".doc", ".docx", ".xls", ".xlsx", ".ppt", ".pptx"),
        "Image Files", Set.of(".jpg", ".jpeg", ".png", ".gif", ".bmp", ".tiff", ".svg"),
        "Binary Files", Set.of(".exe", ".dll", ".so", ".dylib", ".bin"),
        "Database Files", Set.of(".db", ".sqlite", ".mdb", ".accdb"),
        "Script Files", Set.of(".sh", ".bat", ".ps1", ".py", ".js", ".sql")
    );
    
    // Extensions that typically benefit from compression
    private static final Set<String> COMPRESSIBLE_EXTENSIONS = Set.of(
        ".txt", ".dat", ".log", ".csv", ".tsv", ".json", ".xml", 
        ".yaml", ".yml", ".sql", ".html", ".css", ".js", ".py"
    );
    
    // Extensions that are typically already compressed
    private static final Set<String> ALREADY_COMPRESSED_EXTENSIONS = Set.of(
        ".zip", ".gz", ".bz2", ".xz", ".7z", ".rar", ".jpg", ".jpeg", 
        ".png", ".gif", ".mp3", ".mp4", ".avi", ".mov", ".pdf"
    );
    
    // Extensions that require special handling
    private static final Set<String> BINARY_EXTENSIONS = Set.of(
        ".exe", ".dll", ".so", ".dylib", ".bin", ".obj", ".lib", 
        ".jpg", ".jpeg", ".png", ".gif", ".bmp", ".ico"
    );
    
    /**
     * Validate file extension format
     */
    public boolean isValidExtension(String extension) {
        if (extension == null || extension.trim().isEmpty()) {
            return false; // Extension is optional, but if provided must be valid
        }
        
        // Must start with dot and have at least one character after
        return extension.matches("^\\.[a-zA-Z0-9]+$") && extension.length() >= 2 && extension.length() <= 10;
    }
    
    /**
     * Normalize file extension (ensure lowercase and starts with dot)
     */
    public String normalizeExtension(String extension) {
        if (extension == null || extension.trim().isEmpty()) {
            return null;
        }
        
        extension = extension.trim().toLowerCase();
        
        // Add dot if missing
        if (!extension.startsWith(".")) {
            extension = "." + extension;
        }
        
        return extension;
    }
    
    /**
     * Extract file extension from filename
     */
    public String extractExtension(String fileName) {
        if (fileName == null || fileName.trim().isEmpty()) {
            return null;
        }
        
        int lastDotIndex = fileName.lastIndexOf('.');
        if (lastDotIndex > 0 && lastDotIndex < fileName.length() - 1) {
            return fileName.substring(lastDotIndex).toLowerCase();
        }
        
        return null; // No extension found
    }
    
    /**
     * Get file category based on extension
     */
    public String getFileCategory(String extension) {
        if (extension == null) {
            return "Unknown";
        }
        
        extension = normalizeExtension(extension);
        
        for (Map.Entry<String, Set<String>> category : EXTENSION_CATEGORIES.entrySet()) {
            if (category.getValue().contains(extension)) {
                return category.getKey();
            }
        }
        
        return "Other";
    }
    
    /**
     * Check if file extension typically benefits from compression
     */
    public boolean shouldCompress(String extension) {
        if (extension == null) {
            return false;
        }
        
        extension = normalizeExtension(extension);
        return COMPRESSIBLE_EXTENSIONS.contains(extension) && 
               !ALREADY_COMPRESSED_EXTENSIONS.contains(extension);
    }
    
    /**
     * Check if file extension indicates binary content
     */
    public boolean isBinaryExtension(String extension) {
        if (extension == null) {
            return false;
        }
        
        extension = normalizeExtension(extension);
        return BINARY_EXTENSIONS.contains(extension);
    }
    
    /**
     * Get recommended FileType based on extension
     */
    public FileType getRecommendedFileType(String extension) {
        if (extension == null) {
            return FileType.COBOL_FLAT_FILE; // Default
        }
        
        extension = normalizeExtension(extension);
        
        switch (extension) {
            case ".csv":
                return FileType.CSV;
            case ".json":
                return FileType.JSON;
            case ".xml":
                return FileType.XML;
            case ".txt":
            case ".dat":
                return FileType.FIXED_WIDTH;
            case ".tsv":
            case ".pipe":
                return FileType.DELIMITED;
            case ".edi":
                return FileType.EDI;
            default:
                if (isBinaryExtension(extension)) {
                    return FileType.BINARY_FILE;
                } else {
                    return FileType.COBOL_FLAT_FILE;
                }
        }
    }
    
    /**
     * Get all supported file extensions grouped by category
     */
    public Map<String, List<String>> getSupportedExtensionsByCategory() {
        return EXTENSION_CATEGORIES.entrySet().stream()
                .collect(Collectors.toMap(
                    Map.Entry::getKey,
                    entry -> new ArrayList<>(entry.getValue())
                ));
    }
    
    /**
     * Get file extension statistics with categorization
     */
    public Map<String, Object> analyzeExtensions(List<String> extensions) {
        Map<String, Long> categoryCounts = new HashMap<>();
        Map<String, Long> extensionCounts = new HashMap<>();
        
        for (String extension : extensions) {
            if (extension != null) {
                String normalized = normalizeExtension(extension);
                String category = getFileCategory(normalized);
                
                categoryCounts.merge(category, 1L, Long::sum);
                extensionCounts.merge(normalized, 1L, Long::sum);
            }
        }
        
        Map<String, Object> analysis = new HashMap<>();
        analysis.put("totalFiles", extensions.size());
        analysis.put("filesWithExtensions", extensions.stream().filter(Objects::nonNull).count());
        analysis.put("filesWithoutExtensions", extensions.stream().filter(Objects::isNull).count());
        analysis.put("categoryCounts", categoryCounts);
        analysis.put("extensionCounts", extensionCounts);
        analysis.put("mostCommonExtension", getMostCommonExtension(extensionCounts));
        analysis.put("compressibleFiles", extensions.stream().filter(this::shouldCompress).count());
        analysis.put("binaryFiles", extensions.stream().filter(this::isBinaryExtension).count());
        
        return analysis;
    }
    
    /**
     * Validate file extension against allowed extensions for a service
     */
    public ValidationResult validateExtensionForService(String extension, Set<String> allowedExtensions, 
                                                       Set<String> blockedExtensions) {
        if (extension == null) {
            return new ValidationResult(true, "No extension provided (optional)");
        }
        
        extension = normalizeExtension(extension);
        
        // Check if extension is blocked
        if (blockedExtensions != null && blockedExtensions.contains(extension)) {
            return new ValidationResult(false, "File extension is blocked: " + extension);
        }
        
        // Check if extension is allowed (if allowlist is defined)
        if (allowedExtensions != null && !allowedExtensions.isEmpty() && 
            !allowedExtensions.contains(extension)) {
            return new ValidationResult(false, "File extension is not allowed: " + extension);
        }
        
        return new ValidationResult(true, "File extension is valid");
    }
    
    private String getMostCommonExtension(Map<String, Long> extensionCounts) {
        return extensionCounts.entrySet().stream()
                .max(Map.Entry.comparingByValue())
                .map(Map.Entry::getKey)
                .orElse(null);
    }
    
    // Result classes
    
    public static class ValidationResult {
        private final boolean valid;
        private final String message;
        
        public ValidationResult(boolean valid, String message) {
            this.valid = valid;
            this.message = message;
        }
        
        public boolean isValid() { return valid; }
        public String getMessage() { return message; }
    }
}

File: file-transfer-frontend/src/services/fileExtensionService.js
import { api } from './api';

/**
 * Service for file extension operations and utilities
 */
export const fileExtensionService = {
    
    /**
     * Get file transfers by file extension
     */
    async getFileTransfersByExtension(tenantId, extension) {
        try {
            const response = await api.get(`/api/v1/file-transfers/extension/${extension}`, {
                params: { tenantId }
            });
            return response.data;
        } catch (error) {
            console.error('Error fetching file transfers by extension:', error);
            throw error;
        }
    },
    
    /**
     * Get distinct file extensions for a tenant
     */
    async getDistinctFileExtensions(tenantId) {
        try {
            const response = await api.get('/api/v1/file-transfers/extensions', {
                params: { tenantId }
            });
            return response.data;
        } catch (error) {
            console.error('Error fetching distinct file extensions:', error);
            throw error;
        }
    },
    
    /**
     * Get file extension statistics for a tenant
     */
    async getFileExtensionStatistics(tenantId) {
        try {
            const response = await api.get('/api/v1/file-transfers/statistics/extensions', {
                params: { tenantId }
            });
            return response.data;
        } catch (error) {
            console.error('Error fetching file extension statistics:', error);
            throw error;
        }
    },
    
    /**
     * Extract file extension from filename
     */
    extractExtension(fileName) {
        if (!fileName || typeof fileName !== 'string') {
            return null;
        }
        
        const lastDotIndex = fileName.lastIndexOf('.');
        if (lastDotIndex > 0 && lastDotIndex < fileName.length - 1) {
            return fileName.substring(lastDotIndex).toLowerCase();
        }
        
        return null;
    },
    
    /**
     * Normalize file extension (ensure lowercase and starts with dot)
     */
    normalizeExtension(extension) {
        if (!extension || typeof extension !== 'string') {
            return null;
        }
        
        extension = extension.trim().toLowerCase();
        
        // Add dot if missing
        if (!extension.startsWith('.')) {
            extension = '.' + extension;
        }
        
        return extension;
    },
    
    /**
     * Get file category based on extension
     */
    getFileCategory(extension) {
        if (!extension) {
            return 'Unknown';
        }
        
        extension = this.normalizeExtension(extension);
        
        const categories = {
            'Text Files': ['.txt', '.dat', '.log', '.csv', '.tsv'],
            'Data Files': ['.json', '.xml', '.yaml', '.yml', '.properties'],
            'Archive Files': ['.zip', '.gz', '.tar', '.bz2', '.xz', '.7z'],
            'Document Files': ['.pdf', '.doc', '.docx', '.xls', '.xlsx', '.ppt', '.pptx'],
            'Image Files': ['.jpg', '.jpeg', '.png', '.gif', '.bmp', '.tiff', '.svg'],
            'Binary Files': ['.exe', '.dll', '.so', '.dylib', '.bin'],
            'Database Files': ['.db', '.sqlite', '.mdb', '.accdb'],
            'Script Files': ['.sh', '.bat', '.ps1', '.py', '.js', '.sql']
        };
        
        for (const [category, extensions] of Object.entries(categories)) {
            if (extensions.includes(extension)) {
                return category;
            }
        }
        
        return 'Other';
    },
    
    /**
     * Get file extension icon for UI display
     */
    getExtensionIcon(extension) {
        if (!extension) {
            return 'description';
        }
        
        extension = this.normalizeExtension(extension);
        
        const iconMap = {
            '.txt': 'description',
            '.csv': 'table_chart',
            '.json': 'code',
            '.xml': 'code',
            '.pdf': 'picture_as_pdf',
            '.zip': 'folder_zip',
            '.gz': 'folder_zip',
            '.tar': 'folder_zip',
            '.jpg': 'image',
            '.jpeg': 'image',
            '.png': 'image',
            '.gif': 'image',
            '.sql': 'storage',
            '.log': 'article',
            '.dat': 'data_object',
            '.exe': 'apps',
            '.dll': 'extension',
            '.sh': 'terminal',
            '.bat': 'terminal',
            '.py': 'code',
            '.js': 'code',
            '.html': 'web',
            '.css': 'palette'
        };
        
        return iconMap[extension] || 'insert_drive_file';
    },
    
    /**
     * Get file extension color for UI display
     */
    getExtensionColor(extension) {
        if (!extension) {
            return 'default';
        }
        
        const category = this.getFileCategory(extension);
        
        const colorMap = {
            'Text Files': 'primary',
            'Data Files': 'secondary',
            'Archive Files': 'warning',
            'Document Files': 'info',
            'Image Files': 'success',
            'Binary Files': 'error',
            'Database Files': 'primary',
            'Script Files': 'secondary',
            'Other': 'default',
            'Unknown': 'default'
        };
        
        return colorMap[category] || 'default';
    },
    
    /**
     * Check if file extension should be compressed
     */
    shouldCompress(extension) {
        if (!extension) {
            return false;
        }
        
        extension = this.normalizeExtension(extension);
        
        const compressibleExtensions = new Set([
            '.txt', '.dat', '.log', '.csv', '.tsv', '.json', '.xml', 
            '.yaml', '.yml', '.sql', '.html', '.css', '.js', '.py'
        ]);
        
        const alreadyCompressedExtensions = new Set([
            '.zip', '.gz', '.bz2', '.xz', '.7z', '.rar', '.jpg', '.jpeg', 
            '.png', '.gif', '.mp3', '.mp4', '.avi', '.mov', '.pdf'
        ]);
        
        return compressibleExtensions.has(extension) && !alreadyCompressedExtensions.has(extension);
    },
    
    /**
     * Check if file extension indicates binary content
     */
    isBinaryExtension(extension) {
        if (!extension) {
            return false;
        }
        
        extension = this.normalizeExtension(extension);
        
        const binaryExtensions = new Set([
            '.exe', '.dll', '.so', '.dylib', '.bin', '.obj', '.lib', 
            '.jpg', '.jpeg', '.png', '.gif', '.bmp', '.ico', '.pdf',
            '.zip', '.gz', '.bz2', '.xz', '.7z', '.rar'
        ]);
        
        return binaryExtensions.has(extension);
    },
    
    /**
     * Validate file extension format
     */
    validateExtension(extension) {
        const errors = [];
        
        if (extension === null || extension === undefined) {
            return { isValid: true, errors: [], message: 'Extension is optional' };
        }
        
        if (typeof extension !== 'string') {
            errors.push('Extension must be a string');
            return { isValid: false, errors };
        }
        
        extension = extension.trim();
        
        if (extension === '') {
            return { isValid: true, errors: [], message: 'Empty extension is valid (optional)' };
        }
        
        if (!extension.startsWith('.')) {
            errors.push('Extension must start with a dot (.)');
        }
        
        if (extension.length < 2) {
            errors.push('Extension must have at least one character after the dot');
        }
        
        if (extension.length > 10) {
            errors.push('Extension cannot be longer than 10 characters');
        }
        
        if (!/^\.?[a-zA-Z0-9]+$/.test(extension)) {
            errors.push('Extension can only contain letters and numbers');
        }
        
        return {
            isValid: errors.length === 0,
            errors,
            message: errors.length === 0 ? 'Extension is valid' : 'Extension validation failed'
        };
    },
    
    /**
     * Get recommended file extensions for a service type
     */
    getRecommendedExtensions(serviceType) {
        const recommendations = {
            'CUSTOMER_DATA': ['.csv', '.txt', '.dat'],
            'TRANSACTION_DATA': ['.json', '.xml', '.csv'],
            'LOG_DATA': ['.log', '.txt'],
            'CONFIGURATION_DATA': ['.yaml', '.yml', '.properties', '.json'],
            'REPORT_DATA': ['.pdf', '.csv', '.xlsx'],
            'BACKUP_DATA': ['.zip', '.gz', '.tar'],
            'ARCHIVE_DATA': ['.zip', '.7z', '.tar.gz']
        };
        
        return recommendations[serviceType] || ['.txt', '.dat', '.csv', '.json'];
    },
    
    /**
     * Format extension statistics for display
     */
    formatExtensionStatistics(statistics) {
        const total = Object.values(statistics).reduce((sum, count) => sum + count, 0);
        
        return Object.entries(statistics)
            .map(([extension, count]) => ({
                extension,
                count,
                percentage: total > 0 ? Math.round((count / total) * 100 * 100) / 100 : 0,
                category: this.getFileCategory(extension),
                icon: this.getExtensionIcon(extension),
                color: this.getExtensionColor(extension),
                shouldCompress: this.shouldCompress(extension),
                isBinary: this.isBinaryExtension(extension)
            }))
            .sort((a, b) => b.count - a.count);
    },
    
    /**
     * Group extensions by category for display
     */
    groupExtensionsByCategory(extensions) {
        const grouped = {};
        
        extensions.forEach(ext => {
            const category = this.getFileCategory(ext);
            if (!grouped[category]) {
                grouped[category] = [];
            }
            grouped[category].push(ext);
        });
        
        // Sort extensions within each category
        Object.keys(grouped).forEach(category => {
            grouped[category].sort();
        });
        
        return grouped;
    }
};

export default fileExtensionService;